### PR TITLE
Story 302: Severity category work

### PR DESCRIPTION
**Pull request by jsight**
Not quite ready to review yet from a code standpoint, but it should be feature complete now.

### CODE REVIEW

**Comment by jsight** [comment]
retest

**Comment by jsight** [comment]
Related: https://github.com/windup/windup-rulesets/pull/177

File: reporting/api/src/main/java/org/jboss/windup/reporting/severity/IssueCategoryRegistry.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+    /**
+     * Adds a {@link IssueCategory} to the registry and throws a {@link DuplicateIssueCategoryException} if the item already exists.
+     */
+    public void addCategory(IssueCategory category) throws DuplicateIssueCategoryException
+    {
+        IssueCategory original = this.issueCategories.get(category.getCategoryID());
+
+        // Just overwrite it if it was a placeholder
+        if (!original.isPlaceholder())
```
Comment (d-s) [inline reporting/api/src/main/java/org/jboss/windup/reporting/severity/IssueCategoryRegistry.java]: I think you get NPE here once you start adding categories to your XML registry that aren't in the defaults you're adding as placeholders?

Comment (jsight) [inline reporting/api/src/main/java/org/jboss/windup/reporting/severity/IssueCategoryRegistry.java]: Nice catch... fixed (and test-case added). :)

**Comment by jsight** [comment]
@d-s - The pom isn't including the issue descriptions, and that is causing the order to be wrong on the report index. I will look into it for the application list as well.

File: reporting/api/src/main/java/org/jboss/windup/reporting/config/classification/Classification.java [hunk 2]
```diff
@@ -49,20 +51,10 @@
...
-    {
-        return quickfixes;
-    }
-
     private RegexParameterizedPatternParser classificationPattern;
     private RegexParameterizedPatternParser descriptionPattern;
     private int effort;
-    private Severity severity = EffortReportModel.DEFAULT_SEVERITY;
-
+    private IssueCategory issueCategory = null;
```
Comment (mareknovotny) [inline reporting/api/src/main/java/org/jboss/windup/reporting/config/classification/Classification.java]: what about default severity/category ? Some hints/classification uses default setting. Are there migrated to explicit values in the following PR in windup-rulesets?

File: reporting/api/src/main/java/org/jboss/windup/reporting/severity/DuplicateIssueCategoryException.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package org.jboss.windup.reporting.severity;
```
Comment (mareknovotny) [inline reporting/api/src/main/java/org/jboss/windup/reporting/severity/DuplicateIssueCategoryException.java]: is *severity* now category, so why we have package as `org.jboss.windup.reporting.severity` and not `org.jboss.windup.reporting.category`?

File: config-xml/schema/windup-jboss-ruleset.xsd [hunk 4]
```diff
@@ -282,19 +282,15 @@
                 <xs:element ref="quickfix" minOccurs="0" maxOccurs="unbounded"/>
             </xs:sequence>
             <xs:attribute name="title" type="xs:string" use="required"/>
-            <xs:attribute name="severity" type="severity-type" use="optional" />
+            <xs:attribute name="categoryID" type="category-id-type" use="optional" />
```
Comment (jsight) [inline config-xml/schema/windup-jboss-ruleset.xsd]: I know that is true for element names, but I wasn't sure if we wanted the same convention for attribute ids. Do you think "category-id" would be better?

File: reporting/api/src/main/java/org/jboss/windup/reporting/config/classification/Classification.java (hunk 2 above)
Comment (jsight) [inline reporting/api/src/main/java/org/jboss/windup/reporting/config/classification/Classification.java]: It is still setting the category to a default, it is just doing it farther down in the code now.

File: reporting/api/src/main/java/org/jboss/windup/reporting/severity/DuplicateIssueCategoryException.java (hunk 3 above)
Comment (jsight) [inline reporting/api/src/main/java/org/jboss/windup/reporting/severity/DuplicateIssueCategoryException.java]: Good question. I will change it to category.

File: config-xml/schema/windup-jboss-ruleset.xsd (hunk 4 above)
Comment (mareknovotny) [inline config-xml/schema/windup-jboss-ruleset.xsd]: We uses already capital case like afterMethod. So i think going with lowercase or capitalcase is what we have - categoryId or category-id -> choose what you think is better ;)

File: reporting/api/src/main/java/org/jboss/windup/reporting/config/classification/Classification.java (hunk 2 above)
Comment (mareknovotny) [inline reporting/api/src/main/java/org/jboss/windup/reporting/config/classification/Classification.java]: ok, good, I see it now in IssueCategoryRegistry ;)

**Comment by jsight** [comment]
@mareknovotny - Changes made as requested... I think this is ready to review again.

**Comment by mareknovotny** [comment]
great, it looks fine

**Comment by jsight** [comment]
@mareknovotny @johnsteele - Will it cause any issues with the Eclipse tooling to merge it now?

It will require some changes if you are currently tracking master, but probably not if you are on 3.0.0.Alpha1.

**Pull request merged**